### PR TITLE
feat: leetcode zigzag solution

### PR DESCRIPTION
**Pull request by straight-into-the-wall**
#### Description of Change

A solution for the ZigZag conversion problem.

#### References

[Hacktoberfest issue](https://github.com/TheAlgorithms/C/issues/249)

[ZigZag Conversion problem on leetcode](https://leetcode.com/problems/zigzag-conversion/)

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Solution to the 6-ZigZag conversion problem was missing. Here is (decent) one.

[ZigZagConversion-SubmissionDetail-LeetCode.pdf](https://github.com/TheAlgorithms/C/files/7398227/ZigZagConversion-SubmissionDetail-LeetCode.pdf)

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C/pull/897"></a>

### CODE REVIEW

**Comment by Panquesito7** [comment]
Still failing. 👀

**Comment by straight-into-the-wall** [comment]
As specified #900 and #901 were alternatives one to the other.
#900 contains the code tested with this PR and also #902.
#901 contains older code which was probably working sometimes ago.
Without further investigations on the history of this workflow, is preferable to merge only the #900 and revert the #901.

**Comment by Panquesito7** [comment]
Still failing. 👀

**Comment by straight-into-the-wall** [comment]
Somehow this PR was again tested with the buggy version of `awesome_workflow.yml`:

image

You can see in the details of the failing check.

image

and then in failing workflow file:

image

Master branch now seems to have the right `.github/workflows/awesome_workflow.yml`... Maybe relaunch the checks ?

File: leetcode/src/6.c [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ * @brief ZigZag Conversion leetcode problem solution.
+ * @details
+ * A decent solution to the ZigZag conversion problem.
+ * Take advantage of the fact that the maximum gap between the chars is 2 times
+ * the depth(the number of rows).
+ * The actual gap between the two first chars of a rows depends on the depth of
+ * the row. The gaps between successives chars on the same row is the complement
+ * of the first gap to the maximum gap.
+ * @author [straight_into_the_wall](https://github.com/straight-into-the-wall)
+ * @see https://leetcode.com/problems/zigzag-conversion/
```
Comment (Panquesito7) [inline leetcode/src/6.c:12]:
```suggestion
```

File: leetcode/src/6.c [hunk 2]
```diff
@@ -0,0 +1,140 @@
+/**
+ * @file
+ * @brief ZigZag Conversion leetcode problem solution.
```
Comment (Panquesito7) [inline leetcode/src/6.c:3]:
```suggestion
 * @brief Implementation of the [ZigZag Conversion](https://leetcode.com/problems/zigzag-conversion/) Leetcode problem
```

Comment (straight-into-the-wall) [reply leetcode/src/6.c:3]: Definitively better! Thanx!

**Comment by Panquesito7** [comment]
I've seen the tests are failing in other PRs with the same issue you had in this PR. Would you like to investigate? Thanks. 🙂

**Comment by straight-into-the-wall** [comment]
> I've seen the tests are failing in other PRs with the same issue you had in this PR. Would you like to investigate? Thanks. slightly_smiling_face

I checked some of the waiting PR but any of them has the same issue that this one. Could you link some references?

**Comment by Panquesito7** [comment]
> > I've seen the tests are failing in other PRs with the same issue you had in this PR. Would you like to investigate? Thanks. slightly_smiling_face
> 
> I checked some of the waiting PR but any of them has the same issue that this one. Could you link some references?

#914

**Review by Panquesito7 — COMMENTED** [review]
A question: there's no `main` function? 🤔

**Comment by straight-into-the-wall** [comment]
> 

added main function.

File: leetcode/src/6.c [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        "epxcaniewqmoxkjwpymqorluxedvywhcoghotpusfgiestckrpaigocfufbubiyrrffmwa"
+        "eeimidfnnzwccpahtkgfnl"
+        "xpdbsqzsjvctwdrwploufdisxgbahuinogzf";
+
+    testZigZag(s04, 472, r04);
+}
+
+int main(void)
+{
+    test();
```
Comment (Panquesito7) [inline leetcode/src/6.c:143]:
```suggestion
    test();  // run self-test implementations
```

File: leetcode/src/6.c [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        "eccbgqfmzjqtlrsppxqiywjobspefujlxnmddurddiyobqfspvcoulcvdrzkmkwlyiqdch"
+        "ghrgytzdnobqcvdeqjystm"
+        "epxcaniewqmoxkjwpymqorluxedvywhcoghotpusfgiestckrpaigocfufbubiyrrffmwa"
+        "eeimidfnnzwccpahtkgfnl"
+        "xpdbsqzsjvctwdrwploufdisxgbahuinogzf";
+
+    testZigZag(s04, 472, r04);
+}
+
+int main(void)
```
Comment (Panquesito7) [inline leetcode/src/6.c:141]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main()
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there! 😄

**Review by Panquesito7 — APPROVED** [review]
LGTM 🚀 Thanks for your contribution! 😄👍

**Review by amino19 — APPROVED** [review]

**Comment by straight-into-the-wall** [comment]
@ayaankhan98 @mishraabhinn @Panquesito7 ... ping? :wink:

**Pull request merged**